From 6743237507389a63c198d337e8c96b93580f4f22 Mon Sep 17 00:00:00 2001
From: Chandrakala Chavva <Chandrakala.Chavva@cavium.com>
Date: Sat, 1 Aug 2020 00:23:07 -0700
Subject: [PATCH 1446/1921] driver: serdes_debugfs: Add new smc call to tune
 serdes

Added new smc call to tune TX serdes settings

Change-Id: I862ee6557bdfc86bc1e0a0edc7094a26a696ecc1
Signed-off-by: Chandrakala Chavva <Chandrakala.Chavva@cavium.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/33226
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/33376
Signed-off-by: Witold Sadowski <wsadowski@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/50025
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 .../marvell/octeontx2-serdes/serdes_debugfs.c | 136 ++++++++++++++++++
 1 file changed, 136 insertions(+)

diff --git a/drivers/soc/marvell/octeontx2-serdes/serdes_debugfs.c b/drivers/soc/marvell/octeontx2-serdes/serdes_debugfs.c
index 2aefded80a87..4de50050af6e 100644
--- a/drivers/soc/marvell/octeontx2-serdes/serdes_debugfs.c
+++ b/drivers/soc/marvell/octeontx2-serdes/serdes_debugfs.c
@@ -29,6 +29,7 @@ MODULE_LICENSE("GPL v2");
 #define OCTEONTX_SERDES_DBG_GET_EYE	0xc2000d05
 #define OCTEONTX_SERDES_DBG_GET_CONF	0xc2000d06
 #define OCTEONTX_SERDES_DBG_PRBS	0xc2000d07
+#define OCTEONTX_SERDES_DBG_SET_TUNE	0xc2000d08
 
 /* This is expected OcteonTX response for SVC UID command */
 static const int octeontx_svc_uuid[] = {
@@ -96,6 +97,15 @@ static struct {
 	char *res;
 } serdes_cmd_data;
 
+static struct {
+	int qlm;
+	int lane;
+	int swing;
+	int pre;
+	int post;
+	char *res;
+} tune_serdes_cmd;
+
 static struct {
 	int qlm;
 	int qlm_lane;
@@ -414,6 +424,117 @@ static const struct file_operations serdes_dbg_settings_fops = {
 	.release	= single_release,
 };
 
+static int tune_serdes_dbg_lane_parse(const char __user *buffer,
+				 size_t count, int *qlm, int *lane,
+				 int *swing, int *pre, int *post)
+{
+	char *cmd_buf, *cmd_buf_tmp, *subtoken;
+	int ec;
+
+	cmd_buf = memdup_user(buffer, count);
+	if (IS_ERR(cmd_buf))
+		return -ENOMEM;
+
+	cmd_buf[count] = '\0';
+
+	cmd_buf_tmp = strchr(cmd_buf, '\n');
+	if (cmd_buf_tmp) {
+		*cmd_buf_tmp = '\0';
+		count = cmd_buf_tmp - cmd_buf + 1;
+	}
+
+	cmd_buf_tmp = cmd_buf;
+	subtoken = strsep(&cmd_buf, " ");
+	ec = subtoken ? kstrtoint(subtoken, 10, qlm) : -EINVAL;
+
+	if (ec < 0) {
+		kfree(cmd_buf_tmp);
+		return ec;
+	}
+
+	subtoken = strsep(&cmd_buf, " ");
+	ec = subtoken ? kstrtoint(subtoken, 10, lane) : -EINVAL;
+
+	if (ec == -EINVAL) {
+		kfree(cmd_buf_tmp);
+		return ec;
+	}
+
+	subtoken = strsep(&cmd_buf, " ");
+	ec = subtoken ? kstrtoint(subtoken, 10, swing) : -EINVAL;
+
+	if (ec == -EINVAL) {
+		kfree(cmd_buf_tmp);
+		return ec;
+	}
+
+	subtoken = strsep(&cmd_buf, " ");
+	ec = subtoken ? kstrtoint(subtoken, 10, pre) : -EINVAL;
+
+	if (ec == -EINVAL) {
+		kfree(cmd_buf_tmp);
+		return ec;
+	}
+
+	subtoken = strsep(&cmd_buf, " ");
+	ec = subtoken ? kstrtoint(subtoken, 10, post) : -EINVAL;
+
+	kfree(cmd_buf_tmp);
+	return ec;
+}
+
+static int tune_serdes_dbg_settings_read_op(struct seq_file *s, void *unused)
+{
+	tune_serdes_cmd.res = '\0';
+
+	seq_printf(s, "%s", tune_serdes_cmd.res);
+
+	return 0;
+}
+
+static ssize_t tune_serdes_dbg_settings_write_op(struct file *filp,
+					    const char __user *buffer,
+					    size_t count, loff_t *ppos)
+{
+	struct arm_smccc_res res;
+	int ec;
+
+	ec = tune_serdes_dbg_lane_parse(buffer, count, &tune_serdes_cmd.qlm,
+			&tune_serdes_cmd.lane, &tune_serdes_cmd.swing,
+			&tune_serdes_cmd.pre, &tune_serdes_cmd.post);
+	if (ec < 0) {
+		pr_info("Usage: echo <qlm> <lane> <swing> <pre> <post> > tunetx\n");
+		return ec;
+	}
+
+	arm_smccc_smc(OCTEONTX_SERDES_DBG_SET_TUNE, tune_serdes_cmd.qlm,
+		tune_serdes_cmd.lane, tune_serdes_cmd.swing,
+		(tune_serdes_cmd.pre << 8) | (tune_serdes_cmd.post & 0xff),
+		0, 0, 0, &res);
+
+	if (res.a0 != SMCCC_RET_SUCCESS) {
+		pr_info("QLM serdes TX settings command failed.\n");
+		return -EIO;
+	}
+
+	return count;
+}
+
+static int tune_serdes_dbg_open_settings(struct inode *inode, struct file *file)
+{
+	return single_open(file, tune_serdes_dbg_settings_read_op,
+			inode->i_private);
+}
+
+static const struct file_operations tune_serdes_dbg_settings_fops = {
+	.owner		= THIS_MODULE,
+	.open		= tune_serdes_dbg_open_settings,
+	.read		= seq_read,
+	.write		= tune_serdes_dbg_settings_write_op,
+	.llseek		= seq_lseek,
+	.release	= single_release,
+};
+
 static int serdes_dbg_prbs_lane_parse(const char __user *buffer,
 				      size_t count, int *qlm,
 				      enum cgx_prbs_cmd *cmd, int *mode,
@@ -682,6 +803,11 @@ static int serdes_dbg_setup_debugfs(void)
 	if (!pfile)
 		goto create_failed;
 
+	pfile = debugfs_create_file("tunetx", 0644, pserdes_root, NULL,
+				    &tune_serdes_dbg_settings_fops);
+	if (!pfile)
+		goto create_failed;
+
 	return 0;
 
 create_failed:
@@ -727,6 +853,10 @@ static int serdes_dbg_init(void)
 	if (!prbs_cmd_data.res)
 		goto serdes_mem_init_failed;
 
+	tune_serdes_cmd.res = ioremap_wc(res.a0, sizeof(tune_serdes_cmd));
+	if (!tune_serdes_cmd.res)
+		goto serdes_mem_init_failed;
+
 	ec = serdes_dbg_setup_debugfs();
 	if (ec)
 		goto serdes_debugfs_failed;
@@ -748,6 +878,9 @@ static int serdes_dbg_init(void)
 	if (prbs_cmd_data.res)
 		iounmap(prbs_cmd_data.res);
 
+	if (tune_serdes_cmd.res)
+		iounmap(tune_serdes_cmd.res);
+
 	return 0;
 }
 
@@ -766,6 +899,9 @@ static void serdes_dbg_exit(void)
 	if (prbs_cmd_data.res)
 		iounmap(prbs_cmd_data.res);
 
+	if (tune_serdes_cmd.res)
+		iounmap(tune_serdes_cmd.res);
+
 	list_for_each_entry_safe(status, n,
 				 &prbs_cmd_data.status_list.list,
 				 list) {
-- 
2.31.1

